source/application/logging.cpp

```diff
@@ -169,11 +122,111 @@ IDbgHelpStackWalkCallback::Action PrintStackTrace(IDbgHelpStackWalkCallback& dat
+}
+
+// TODO: Break filewriting out into it's own globally set function that can be replaced.
+DebugLog::~DebugLog()
+{
+    m_buffer << L"\n";
+    if (IsDebuggerPresent())
+    {
+        OutputDebugStringW(m_buffer.str().c_str());
+    }
+    EnterCriticalSection(&gs_debug_print_cs);
+    if (!gs_debuglogfile.is_open())
+    {
+        gs_debuglogfile.open("hourglasslog.txt", std::wifstream::out);
+    }
+    if (gs_debuglogfile.is_open())
+    {
+        gs_debuglogfile << m_buffer.str();
+        gs_debuglogfile.flush();
+    }
+    LeaveCriticalSection(&gs_debug_print_cs);
+}
+
+DebugLog& DebugLog::operator<<(const IPC::DebugMessage* dbgmsg)
+{
+    /*
+     * Work-around because setw and setfill appends before the '0x' provided by showbase.
+     */
+    m_buffer.unsetf(std::ios::showbase);
+    const BYTE* data = dbgmsg->GetDebugMessage();
+    for (DWORD type = *reinterpret_cast<const DWORD*>(data); type != 0; type = *reinterpret_cast<const DWORD*>(data))
+    {
+        data += sizeof(DWORD);
+        DWORD size = *reinterpret_cast<const DWORD*>(data);
+        data += sizeof(DWORD);
+
+        switch (type)
+        {
+        case 'X':
+        case 'p':
+            {
+                m_buffer << "0x" << std::hex << std::setfill(L'0') << std::setw(size * 2);
+                if (size == sizeof(unsigned char))
+                {
+                    m_buffer << *reinterpret_cast<const unsigned char*>(data);
+                }
+                else if (size == sizeof(unsigned short))
+                {
+                    m_buffer << *reinterpret_cast<const unsigned short*>(data);
+                }
+                else if (size == sizeof(unsigned int))
+                {
+                    m_buffer << *reinterpret_cast<const unsigned int*>(data);
+                }
+                else if (size == sizeof(unsigned long long int))
+                {
+                    m_buffer << *reinterpret_cast<const unsigned long long int*>(data);
+                }
+            }
+            break;
+        case 'g':
+            {
+                m_buffer << std::dec;
+                if (size == sizeof(float))
+                {
+                    m_buffer << *reinterpret_cast<const float*>(data);
+                }
+                else if (size == sizeof(double))
+                {
+                    m_buffer << *reinterpret_cast<const double*>(data);
+                }
+            }
+            break;
+        case 'b':
+            {
+                m_buffer << std::boolalpha << *reinterpret_cast<const bool*>(data);
+            }
+            break;
+        case 'S':
+            {
+                m_buffer << reinterpret_cast<LPCSTR>(data);
+            }
+            break;
+        case 's':
+            {
+                m_buffer << reinterpret_cast<LPCWSTR>(data);
+            }
+            break;
+        default:
+            assert(false);
+            break;
+        }
+
+        data += size + (4 - (size % 4));

```

Comment:
This will do `data += size + 4` if `size == 4` or `size == 0`, while it should do just `data += size`?